File: 프로그래머스/l2_프린터.py
from collections import deque

def solution(prior, location):

    output = []
    priority = deque(prior)
    numbers = deque([i for i in range(len(prior))])

    while priority:
        max_num = max(priority)

        j = priority.popleft()
        n = numbers.popleft()
        if max_num == j:
            output.append(n)
            
        else:
            priority.append(j)
            numbers.append(n)

    return output.index(location)+1




solution([2,1,3,2], 2)
solution([1,1,9,1,1,1],0)


















# import collections
# def solution(prior, location):
#     queue = collections.deque(prior)
#     idx = collections.deque([i for i in range(len(prior))])
#     result = []

#     while queue:
#         v = queue.popleft()
#         i = idx.popleft()
        
#         if not queue or max(queue) <=v:
#             result.append(i)
#         elif max(queue) > v:
#             queue.append(v)
#             idx.append(i)

#     return result.index(location)+1

File: 프로그래머스/JAVA/l2_프린터.java
import java.util.*;

public class l2_프린터 {
    public static void main(String[] agrs){
        int[] properties = {2,1,3,2};
        int location =2;
        int answer = solution(properties, location);
    }
    public static int solution(int[] priorities, int location) {
        Deque<Integer> queue = new ArrayDeque<>();
        Deque<Integer> numbers = new ArrayDeque<>();
        ArrayList<Integer> answer = new ArrayList<>();

        for (int i=0; i<priorities.length; i++){
            queue.add(priorities[i]);
            numbers.add(i);
        }

        while (!queue.isEmpty()){
            int nowPriority = queue.pollFirst();
            int nowIndex = numbers.pollFirst();
            System.out.println(nowIndex + " " + nowPriority);
            System.out.println(queue +" " + numbers);
            for (int priority : queue){
                if (priority > nowPriority){
                    queue.addLast(nowPriority);
                    numbers.addLast(nowIndex);
                    nowIndex = -1;
                    break;
                }
            }
            if (nowIndex >=0)
                answer.add(nowIndex);

        }
        System.out.println(answer);
        System.out.println(answer.indexOf(location));

        return answer.indexOf(location)+1;
    }
}
